### PR TITLE
Refactor get config to `getParameterFromConfigV2` Kafka Scaler

### PR DESCRIPTION
**Pull request by dttung2905**
Hi team,

I did the refactor for 1 scaler first to get your opinion on this. Chose Kafka because imo its the most complexed :sweat: so its worth a try

### Checklist

- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [ ] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Relates to https://github.com/kedacore/keda/issues/5037

### CODE REVIEW

**Review by wozniakjan — COMMENTED** [review]
can you please add a changelog entry? I guess it should go in the `other` section as `Kafka scaler: ...`

File: pkg/scalers/apache_kafka_scaler.go [hunk 1]
```diff
@@ -120,84 +120,148 @@ func NewApacheKafkaScaler(ctx context.Context, config *scalersconfig.ScalerConfi
...
-			return fmt.Errorf("error incorrect TLS value given, got %s", val)
-		}
+	var enableTLS bool
+	tlsString, err := getParameterFromConfigV2(
+		config,
+		"tls",
+		reflect.TypeOf(""),
+		UseMetadata(true),
+		UseAuthentication(true),
+		UseResolvedEnv(false),
```
Comment (wozniakjan) [inline pkg/scalers/apache_kafka_scaler.go:130]: nit: iirc, the default value for optional arg is false, so the `UseResolvedEnv(false)` could be omitted?

File: pkg/scalers/apache_kafka_scaler.go [hunk 2]
```diff
@@ -120,84 +120,148 @@ func NewApacheKafkaScaler(ctx context.Context, config *scalersconfig.ScalerConfi
...
+		IsOptional(true),
+		WithDefaultVal("disable"),
```
Comment (wozniakjan) [inline pkg/scalers/apache_kafka_scaler.go:132]: nit: iirc, the default value for optional arg is false, so the `UseResolvedEnv(false)` could be omitted? and for brevity, can we omit the arguments?

```suggestion
	tlsString, err := getParameterFromConfigV2(
		config,
		"tls",
		reflect.TypeOf(""),
		UseMetadata(),
		UseAuthentication(),
		IsOptional(),
		WithDefaultVal("disable"),
```

Comment (dttung2905) [reply pkg/scalers/apache_kafka_scaler.go:132]: I think eleminating option function when it is not `true` make sense as we already initiate the default value in this 
https://github.com/kedacore/keda/blob/ef464594c311818fb400cc8d69a8d6e9721da21b/pkg/scalers/scaler.go#L243

As fo having the default value, I personally does not enjoy that as Golang does not support default arg value out of the box :laughing: , although we can still make some changes to adopt this, for example 
Before 
```golang
func IsOptional(optional bool) Option {
	return func(opt *configOptions) {
		opt.isOptional = optional
	}
}
```
After
```golang
func IsOptional(optionals ...bool) Option {
    optional := true // Default value
    if len(optionals) > 0 {
        optional = optionals[0]
    }
    return func(opt *configOptions) {
        opt.isOptional = optional
    }
}
```
Let me know if you still want to go through with the second point on default args value

**Comment by wozniakjan** [comment]
I would personally just drop the argument from all `optional` helper functions
```
UseMetadata(true)
UseAuthentication(true)
IsOptional(true)
...
```
would become
```
UseMetadata()
UseAuthentication()
IsOptional()
...
```
the bool argument is imho not necessary and the absence of `option` is semantically equivalent to `option(false)`.

**Comment by dttung2905** [comment]
@JorTurFer Could you help to take a look at this when you have time ?

**Comment by JorTurFer** [comment]
Sorry, I've been travelling this week :( 
Honestly, I love this new approach for parsing parameters. It unifies all the process and it's awesome, great job!

I agree with the suggestions from @wozniakjan , but you've already addressed them, so nothing else from my side 🥇

**Review by JorTurFer — APPROVED** [review]

**Comment by dttung2905** [comment]
let me take a look at the failed e2e tests

**Comment by dttung2905** [comment]
Seems like the similar failures is happening on the daily E2E test on `main` branch too [link](https://github.com/kedacore/keda/actions/runs/8026638253)
image 

The current failed [E2E test](https://github.com/kedacore/keda/actions/runs/8032604027/job/21942265728#step:9:2796) shows similar pattern ( not exactly the same test case but still within the apache kafka scaler test )
The weird thing is when I try to reproduce the E2E tests on my local machine, it is showing time out 
```
 apache_kafka_test.go:610: --- testing  limitToPartitionsWithLag: no scale out ---
    helper.go:548: Applying template: singleDeploymentTemplate
    helper.go:548: Applying template: limitToPartionsWithLagScaledObjectTemplate
    helper.go:500: Waiting for some time to ensure deployment replica count doesn't change from 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:500: Waiting for some time to ensure deployment replica count doesn't change from 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
    helper.go:507: Deployment - apache-kafka-test-deployment, Current  - 0
panic: test timed out after 10m0s
running tests:
	TestScaler (10m0s)

goroutine 1152 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/go/src/testing/testing.go:2259 +0x3b9
created by time.goFunc
	/usr/local/go/src/time/sleep.go:176 +0x2d

goroutine 1 [chan receive, 10 minutes]:
testing.(*T).Run(0xc000304820, {0x186761b?, 0x52c5dc?}, 0x195caf8)
	/usr/local/go/src/testing/testing.go:1649 +0x3c8
testing.runTests.func1(0x26c6e00?)
	/usr/local/go/src/testing/testing.go:2054 +0x3e
testing.tRunner(0xc000304820, 0xc00019fc48)
	/usr/local/go/src/testing/testing.go:1595 +0xff
testing.runTests(0xc000238be0?, {0x26a0640, 0x1, 0x1}, {0x418a05?, 0xc00019fd08?, 0x26c5b20?})
	/usr/local/go/src/testing/testing.go:2052 +0x445
testing.(*M).Run(0xc000238be0)
	/usr/local/go/src/testing/testing.go:1925 +0x636
main.main()
	_testmain.go:47 +0x19c

```
The errors come from this line https://github.com/kedacore/keda/blob/b3f554899d610cc9d7c5f6a8f94b404ce829876d/tests/scalers/apache_kafka/apache_kafka_test.go#L635

@JorTurFer Do you know how can we can increase the timeout to greater than `10m` ? I tried to reduce the wait time of the `AssertReplicaCountNotChangeDuringTimePeriod` from `30` to `3` so the test passes that line but still throw a `Timed out` error down in the later section of the code

**Comment by JorTurFer** [comment]
> Do you know how can we can increase the timeout to greater than 10m ?

Actually, the timeout is 20m:
https://github.com/kedacore/keda/blob/f2d86a887d03454adaa6b91be7dcbfa264422b43/tests/run-all.go#L27-L33

**Review by zroubalik — COMMENTED** [review]
@wozniakjan so are we gonna give this a try?

**Comment by wozniakjan** [comment]
I think there is great value in this PR, it's been reviewed, I vote for merge.

Parallel to this, I'm curious what the feedback will be on https://github.com/kedacore/keda/pull/5676, perhaps that approach can make the code even more readable.

**Comment by JorTurFer** [comment]
hum... it's a fascinating approach that one which you're proposing tbh. It also can help decoupling the scaling metadata parsing from the scaler itself based on the tagging model and it's also really interesting IMHO.
WDYT @tomkerkhove  @dttung2905 @zroubalik ?

**Comment by dttung2905** [comment]
> hum... it's a fascinating approach that one which you're proposing tbh. It also can help decoupling the scaling metadata parsing from the scaler itself based on the tagging model and it's also really interesting IMHO. WDYT @tomkerkhove @dttung2905 @zroubalik ?

I think this is an interesting approach from @wozniakjan , we should definitely give it a try. We can just keep this PR open as a reference in case we want to come back to this in the near future

**Comment by wozniakjan** [comment]
thank you, I will add https://github.com/kedacore/keda/pull/5676 to the agenda for the next KEDA community meeting, so we can discuss the advantages and disadvantages of the declarative config approach.

**Pull request closed without merging**